rpcs3/Emu/RSX/RSXThread.cpp

```diff
@@ -2483,14 +2478,36 @@ namespace rsx
 
 			_mm_pause();
 		}
-
-		external_interrupt_ack.store(false);
 	}
 
 	void thread::unpause()
 	{
 		// TODO: Clean this shit up
-		external_interrupt_lock.store(false);
+		if (external_interrupt_lock-- == 1)
+		{
+			external_interrupt_lock.notify_one();
+		}
+	}
+
+	void thread::wait_pause()
+	{
+		do
+		{
+			if (g_cfg.video.multithreaded_rsx)
+			{
+				g_dma_manager.sync();
+			}
+
+			external_interrupt_ack.store(true);
+
+			while (const u32 old = external_interrupt_lock)
+			{
+				external_interrupt_lock.wait(old);
```

Comment:
Hmm, last time I profiled this it wasn't that great for super-short locks on windows, but that was many months ago so things may have changed. Remember pause/unpause is also used in sys_rsx whenever anything needs updating, sometimes just to wait for 2 int values to be changed so it should not have measurable lag.